src/audio_decoder_midi.h

```diff
@@ -106,45 +146,66 @@ class GenericMidiDecoder : public AudioDecoder, midisequencer::output {
 	 */
 	int GetTicks() const override;
 
+	/**
+	 * Generate a MIDI reset event so the device doesn't
+	 * leave notes playing or keeps any state.
+	 */
+	void Reset();
+
 	std::vector<uint8_t> file_buffer;
 	size_t file_buffer_pos = 0;
 private:
 	static constexpr int midi_default_tempo = 500000;
 
-	std::unique_ptr<MidiDecoder> mididec;
-
 	int FillBuffer(uint8_t* buffer, int length) override;
 
-	float mtime = 0.0f;
+	void SendMessageToAllChannels(uint32_t midi_msg);
+
+	// midisequencer::output interface
+	void midi_message(int, uint_least32_t message) override;
+	void sysex_message(int, const void* data, std::size_t size) override;
+	void meta_event(int, const void*, std::size_t) override;
+	void reset() override;
+	void reset_tempos_after_loop();
+
+	std::chrono::microseconds mtime = std::chrono::microseconds(0);
 	float pitch = 1.0f;
-	int frequency = 44100;
+	bool paused = false;
+	float volume = 0;
 	bool loops_to_end = false;
 
+	int fade_steps = 0;
+	double fade_volume_end = 0;
+	double delta_volume_step = 0;
+	std::chrono::microseconds last_fade_mtime = std::chrono::microseconds(0);
+
+	int frequency = 44100;
```

Comment:
Could this use EP_MIDI_FREQUENCY or is it unrelated?

Reply:
good find. This must be ``EP_MIDI_FREQUENCY``

Reply:
actually this value is not used, at least when the caller is well-behaving and does "SetFormat"